### PR TITLE
Call setxkbmap when changing the keyboard layout selection.

### LINKED ISSUE #148: Selecting a keyboard layout has no effect
Selecting a keyboard layout in the list doesn't change the active one, that's because the setKeyboard API from subiquity calls setupcon which works for VTs but not graphical sessions

The standard way to change the active layout for Xorg is to call setxkbmap -layout ... -variant ...

Under wayland there is no equivalent though. One way could be to call to localectl set-keymap and to teach g-s-d to listen to localed changes and reflect them?

### PR DESCRIPTION
**Pull request by oSoMoN**
Fixes #148.

### CODE REVIEW

**Comment by seb128** [comment]
Thanks Olivier, I tested and from an user perspective it is doing the work on an xorg session. One nitpick, the code doesn't handle cleanly setxkbmap not being installed and throw exception on stderr, would be nicer to rather log a warning or just do nothing in this case (the utility is standard but it's unclear it will always be available in a wayland world since it doesn't work there)

**Comment by mwhudson** [comment]
I was sort of thinking we'd add code to the subiquity side for this, but I guess it's not required...

**Comment by oSoMoN** [comment]
> I was sort of thinking we'd add code to the subiquity side for this, but I guess it's not required...

If calling `setxkbmap` has any benefit in a server environment, then sure, let's implement it in subiquity. Otherwise it's probably fine to have that logic remain in the desktop installer.

**Comment by seb128** [comment]
> I was sort of thinking we'd add code to the subiquity side for this, but I guess it's not required...

It would perhaps still be useful if subiquity was using 'localectl set-keymap ...' rather than setupcon? we will eventually need to solve the wayland case and using localed might be the right solution

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart [hunk 1]
```diff
@@ -39,24 +47,22 @@ class _KeyboardLayoutPageState extends State<KeyboardLayoutPage> {
           return layout.code == info.layout;
         });
         if (_selectedLayoutIndex > -1) {
-          _selectedLayoutName =
-              keyboardModel.layouts[_selectedLayoutIndex].code!;
-          _selectedVariantIndex = keyboardModel
-              .layouts[_selectedLayoutIndex].variants!
+          _selectedVariantIndex = _selectedLayout!.variants!
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart:50]: is the list of variants guaranteed to be non-null or should it be `.variants?.indexWhere(...) ?? -1`?

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart [hunk 2]
```diff
@@ -167,9 +181,8 @@ class _KeyboardLayoutPageState extends State<KeyboardLayoutPage> {
                 final client =
                     Provider.of<SubiquityClient>(context, listen: false);
                 final keyboard = KeyboardSetting(
-                    layout: _selectedLayoutName,
-                    variant: keyboardModel.layouts[_selectedLayoutIndex]
-                        .variants![_selectedVariantIndex].code!);
+                    layout: _selectedLayout!.code!,
+                    variant: _selectedVariant!.code!);
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart:185]: - should the wizard action be disabled if there's no selected layout?
- KeyboardLayout.variants is nullable => can the selected variant be null at this point?

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart (hunk 1 above)
Comment (oSoMoN) [reply packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart]: You're right, of course. Fixed.

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart (hunk 2 above)
Comment (oSoMoN) [reply packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart:188]: Very good point. I disabled the button if there's no layout/variant selected. This ensures that when the action is triggered, the selected variant isn't null.

**Review by jpnurmi — APPROVED** [review]
LGTM :+1:

**Comment by mwhudson** [comment]
Thinking about this a little more, I think doing this in the client for now makes total sense. If keyboard setting in general gets unified a bit we can look at making the server part use that.

File: packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart [hunk 3]
```diff
@@ -47,19 +49,20 @@ class _KeyboardLayoutPageState extends State<KeyboardLayoutPage> {
         if (_selectedLayoutIndex > -1) {
           _selectedVariantIndex = _selectedLayout!.variants!
               .indexWhere((variant) => variant.code == info.variant);
-          if (_selectedVariantIndex == -1) {
-            _selectedVariantIndex = 0;
```
Comment (jpnurmi) [inline packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart:48]: is it intentional that the default variant is no longer selected? the above `indexWhere` is comparing empty to null.

Screenshot from 2021-07-14 07-44-00

Comment (jpnurmi) [reply packages/ubuntu_desktop_installer/lib/pages/keyboard_layout_page.dart:48]: Fix proposal: https://github.com/canonical/ubuntu-desktop-installer/pull/168

**Comment by jpnurmi** [comment]
I noticed that `setxkbmap` is currently called in dry-run mode (https://github.com/canonical/ubuntu-desktop-installer/issues/250). Could we do it in subiquity so the UI doesn't need to know if it's running in dry-run mode?

**Pull request merged**